File: packages/lenso-process-protocol/src/proof.ts
import type { HandshakeParams } from "./types.js";
import { validateInitialize, type InitializeParams } from "./authoring.js";

const encoder = new TextEncoder();
const HOST_DOMAIN = encoder.encode("lenso-process-host-v1");
const CHILD_DOMAIN = encoder.encode("lenso-process-child-v1");
const AUTHORING_HOST_DOMAIN = encoder.encode("lenso-authoring-host-v2");
const AUTHORING_CHILD_DOMAIN = encoder.encode("lenso-authoring-child-v2");
const AUTHORING_CALLBACK_DOMAIN = encoder.encode("lenso-authoring-callback-v2");
const BASE64URL = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

/** RFC 8785-compatible canonical JSON for the restricted proof value profile. */
export function canonicalizeProofValue(value: unknown): Uint8Array {
  return encoder.encode(canonical(value));
}

/** Canonical bytes hashed to obtain `handshake_params_digest`. */
export function handshakeProofPayload(params: HandshakeParams): Uint8Array {
  return canonicalizeProofValue({
    identity: params.identity,
    host_nonce: params.host_nonce,
  });
}

/** Exact bytes authenticated by `host_proof`. */
export function hostProofMessage(handshakeDigest: Uint8Array): Uint8Array {
  exactLength(handshakeDigest, 32, "handshake digest");
  return concatenate(HOST_DOMAIN, Uint8Array.of(0), handshakeDigest);
}

/** Exact bytes authenticated by `child_proof`. */
export function childProofMessage(
  handshakeDigest: Uint8Array,
  session: string,
): Uint8Array {
  exactLength(handshakeDigest, 32, "handshake digest");
  return concatenate(
    CHILD_DOMAIN,
    Uint8Array.of(0),
    handshakeDigest,
    decodeBase64Url32(session, "session"),
  );
}

export interface AuthoringHandshakeProofInput {
  readonly initialize: InitializeParams;
  readonly callback_origin: string;
  readonly host_nonce: string;
}

/** Canonical bytes hashed before authenticating one Authoring V2 initialization. */
export function authoringHandshakeProofPayload(
  input: AuthoringHandshakeProofInput,
): Uint8Array {
  validateInitialize(input.initialize);
  if (!isLoopbackHttpOrigin(input.callback_origin)) {
    throw new Error("callback_origin must be an exact loopback HTTP origin");
  }
  decodeBase64Url32(input.host_nonce, "host_nonce");
  return canonicalizeProofValue(input);
}

/** Exact bytes authenticated by the Authoring V2 Host proof. */
export function authoringHostProofMessage(handshakeDigest: Uint8Array): Uint8Array {
  exactLength(handshakeDigest, 32, "handshake digest");
  return concatenate(AUTHORING_HOST_DOMAIN, Uint8Array.of(0), handshakeDigest);
}

/** Exact bytes authenticated by the Authoring V2 child proof. */
export function authoringChildProofMessage(
  handshakeDigest: Uint8Array,
  childNonce: string,
): Uint8Array {
  exactLength(handshakeDigest, 32, "handshake digest");
  return concatenate(
    AUTHORING_CHILD_DOMAIN,
    Uint8Array.of(0),
    handshakeDigest,
    decodeBase64Url32(childNonce, "child_nonce"),
  );
}

/** Exact bytes authenticating one child-to-Host callback request. */
export function authoringCallbackProofMessage(
  session: string,
  method: "lenso.call" | "lenso.settled",
  params: unknown,
): Uint8Array {
  const sessionBytes = decodeBase64Url32(session, "session");
  return concatenate(
    AUTHORING_CALLBACK_DOMAIN,
    Uint8Array.of(0),
    sessionBytes,
    Uint8Array.of(0),
    encoder.encode(method),
    Uint8Array.of(0),
    canonicalizeProofValue(params),
  );
}

/** Decodes one canonical unpadded base64url value containing exactly 32 bytes. */
export function decodeBase64Url32(value: string, name = "value"): Uint8Array {
  if (!/^[A-Za-z0-9_-]{43}$/.test(value)) {
    throw new Error(`${name} must encode exactly 32 canonical base64url bytes`);
  }
  const decoded = decodeBase64Url(value);
  exactLength(decoded, 32, name);
  if (encodeBase64Url(decoded) !== value) {
    throw new Error(`${name} must use canonical unpadded base64url`);
  }
  return decoded;
}

/** Encodes bytes as canonical unpadded base64url. */
export function encodeBase64Url(value: Uint8Array): string {
  let output = "";
  for (let index = 0; index < value.length; index += 3) {
    const first = value[index] as number;
    const second = value[index + 1];
    const third = value[index + 2];
    const bits = (first << 16) | ((second ?? 0) << 8) | (third ?? 0);
    output += BASE64URL[(bits >>> 18) & 63];
    output += BASE64URL[(bits >>> 12) & 63];
    if (second !== undefined) output += BASE64URL[(bits >>> 6) & 63];
    if (third !== undefined) output += BASE64URL[bits & 63];
  }
  return output;
}

function decodeBase64Url(value: string): Uint8Array {
  let buffer = 0;
  let bits = 0;
  const output: number[] = [];
  for (const character of value) {
    const digit = BASE64URL.indexOf(character);
    if (digit < 0) throw new Error("invalid base64url character");
    buffer = (buffer << 6) | digit;
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      output.push((buffer >>> bits) & 0xff);
      buffer &= (1 << bits) - 1;
    }
  }
  if (bits > 0 && buffer !== 0) throw new Error("non-canonical base64url tail bits");
  return Uint8Array.from(output);
}

function canonical(value: unknown): string {
  if (value === null) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "string") return JSON.stringify(value);
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) {
      throw new Error("proof JSON permits only portable safe integers");
    }
    return String(value);
  }
  if (Array.isArray(value)) return `[${value.map(canonical).join(",")}]`;
  if (typeof value === "object") {
    const candidate = value as Record<string, unknown>;
    const entries = Object.keys(candidate)
      .sort()
      .map((key) => `${JSON.stringify(key)}:${canonical(candidate[key])}`);
    return `{${entries.join(",")}}`;
  }
  throw new Error("proof JSON contains an unsupported value");
}

function concatenate(...values: readonly Uint8Array[]): Uint8Array {
  const output = new Uint8Array(values.reduce((length, value) => length + value.length, 0));
  let offset = 0;
  for (const value of values) {
    output.set(value, offset);
    offset += value.length;
  }
  return output;
}

function exactLength(value: Uint8Array, expected: number, name: string): void {
  if (value.length !== expected) throw new Error(`${name} must contain ${expected} bytes`);
}

function isLoopbackHttpOrigin(value: string): boolean {
  const match = /^http:\/\/(?:127\.0\.0\.1|\[::1\]):([1-9][0-9]{0,4})\/$/.exec(value);
  return match !== null && Number(match[1]) <= 65_535;
}

File: crates/lenso-process-protocol/src/proof.rs
use serde::Serialize;
use serde_json::Value;
use sha2::{Digest as _, Sha256};

use super::{
    HandshakeIdentity, HandshakeParams, ProtocolError, authoring::InitializeParams,
    decode_base64url_32,
};

const HOST_PROOF_DOMAIN: &[u8] = b"lenso-process-host-v1";
const CHILD_PROOF_DOMAIN: &[u8] = b"lenso-process-child-v1";
const AUTHORING_HOST_PROOF_DOMAIN: &[u8] = b"lenso-authoring-host-v2";
const AUTHORING_CHILD_PROOF_DOMAIN: &[u8] = b"lenso-authoring-child-v2";
const AUTHORING_CALLBACK_PROOF_DOMAIN: &[u8] = b"lenso-authoring-callback-v2";
const MAX_SAFE_JSON_INTEGER: u64 = 9_007_199_254_740_991;

/// Returns RFC 8785-compatible canonical bytes for the restricted proof profile.
pub fn canonicalize_proof_value(value: &Value) -> Result<Vec<u8>, ProtocolError> {
    let mut output = Vec::new();
    write_canonical(value, &mut output)?;
    Ok(output)
}

/// Computes the raw SHA-256 digest bound by host and child handshake proofs.
pub fn handshake_params_digest(params: &HandshakeParams) -> Result<[u8; 32], ProtocolError> {
    #[derive(Serialize)]
    struct ProofPayload<'a> {
        identity: &'a HandshakeIdentity,
        host_nonce: &'a str,
    }

    params.identity.validate()?;
    decode_base64url_32("host_nonce", &params.host_nonce)?;
    let value = serde_json::to_value(ProofPayload {
        identity: &params.identity,
        host_nonce: &params.host_nonce,
    })
    .map_err(|error| ProtocolError::new(format!("cannot encode handshake proof: {error}")))?;
    let canonical = canonicalize_proof_value(&value)?;
    Ok(Sha256::digest(canonical).into())
}

/// Returns the exact bytes authenticated by `host_proof`.
#[must_use]
pub fn host_proof_message(handshake_digest: &[u8; 32]) -> Vec<u8> {
    proof_message(HOST_PROOF_DOMAIN, handshake_digest, None)
}

/// Returns the exact bytes authenticated by `child_proof`.
pub fn child_proof_message(
    handshake_digest: &[u8; 32],
    session: &str,
) -> Result<Vec<u8>, ProtocolError> {
    let session = decode_base64url_32("session", session)?;
    Ok(proof_message(
        CHILD_PROOF_DOMAIN,
        handshake_digest,
        Some(&session),
    ))
}

/// Values bound by an Authoring V2 initialization handshake.
#[derive(Clone, Copy, Debug)]
pub struct AuthoringHandshakeProofInput<'a> {
    /// Exact initialization admitted by the child.
    pub initialize: &'a InitializeParams,
    /// Exact loopback HTTP origin exposed by the Host callback listener.
    pub callback_origin: &'a str,
    /// Canonical 32-byte base64url Host nonce.
    pub host_nonce: &'a str,
}

/// Returns the canonical bytes hashed before authenticating Authoring V2 initialization.
pub fn authoring_handshake_proof_payload(
    input: AuthoringHandshakeProofInput<'_>,
) -> Result<Vec<u8>, ProtocolError> {
    #[derive(Serialize)]
    struct ProofPayload<'a> {
        initialize: &'a InitializeParams,
        callback_origin: &'a str,
        host_nonce: &'a str,
    }

    input.initialize.validate()?;
    validate_loopback_http_origin(input.callback_origin)?;
    decode_base64url_32("host_nonce", input.host_nonce)?;
    let value = serde_json::to_value(ProofPayload {
        initialize: input.initialize,
        callback_origin: input.callback_origin,
        host_nonce: input.host_nonce,
    })
    .map_err(|error| ProtocolError::new(format!("cannot encode authoring proof: {error}")))?;
    canonicalize_proof_value(&value)
}

/// Returns the exact bytes authenticated by the Authoring V2 Host proof.
#[must_use]
pub fn authoring_host_proof_message(handshake_digest: &[u8; 32]) -> Vec<u8> {
    proof_message(AUTHORING_HOST_PROOF_DOMAIN, handshake_digest, None)
}

/// Returns the exact bytes authenticated by the Authoring V2 child proof.
pub fn authoring_child_proof_message(
    handshake_digest: &[u8; 32],
    child_nonce: &str,
) -> Result<Vec<u8>, ProtocolError> {
    let child_nonce = decode_base64url_32("child_nonce", child_nonce)?;
    Ok(proof_message(
        AUTHORING_CHILD_PROOF_DOMAIN,
        handshake_digest,
        Some(&child_nonce),
    ))
}

/// Returns the exact bytes authenticating one child-to-Host callback request.
pub fn authoring_callback_proof_message(
    session: &str,
    method: &str,
    params: &Value,
) -> Result<Vec<u8>, ProtocolError> {
    if !matches!(
        method,
        "lenso.call"
            | "lenso.event.publish"
            | "lenso.stream.open"
            | "lenso.stream.send"
            | "lenso.stream.receive"
            | "lenso.stream.close_send"
            | "lenso.stream.cancel"
            | "lenso.settled"
    ) {
        return Err(ProtocolError::new("unsupported Authoring callback method"));
    }
    let session = decode_base64url_32("session", session)?;
    let canonical_params = canonicalize_proof_value(params)?;
    let mut message = Vec::with_capacity(
        AUTHORING_CALLBACK_PROOF_DOMAIN.len()
            + 1
            + session.len()
            + 1
            + method.len()
            + 1
            + canonical_params.len(),
    );
    message.extend_from_slice(AUTHORING_CALLBACK_PROOF_DOMAIN);
    message.push(0);
    message.extend_from_slice(&session);
    message.push(0);
    message.extend_from_slice(method.as_bytes());
    message.push(0);
    message.extend_from_slice(&canonical_params);
    Ok(message)
}

fn validate_loopback_http_origin(value: &str) -> Result<(), ProtocolError> {
    let port = value
        .strip_prefix("http://127.0.0.1:")
        .or_else(|| value.strip_prefix("http://[::1]:"))
        .and_then(|value| value.strip_suffix('/'))
        .ok_or_else(|| {
            ProtocolError::new("callback_origin must be an exact loopback HTTP origin")
        })?;
    if !port.as_bytes().first().is_some_and(u8::is_ascii_digit)
        || port.starts_with('0')
        || port
            .parse::<u16>()
            .ok()
            .as_ref()
            .is_none_or(|port| *port == 0)
    {
        return Err(ProtocolError::new(
            "callback_origin must be an exact loopback HTTP origin",
        ));
    }
    Ok(())
}

fn proof_message(domain: &[u8], digest: &[u8; 32], suffix: Option<&[u8; 32]>) -> Vec<u8> {
    let mut message = Vec::with_capacity(domain.len() + 1 + 32 + suffix.map_or(0, |_| 32));
    message.extend_from_slice(domain);
    message.push(0);
    message.extend_from_slice(digest);
    if let Some(suffix) = suffix {
        message.extend_from_slice(suffix);
    }
    message
}

fn write_canonical(value: &Value, output: &mut Vec<u8>) -> Result<(), ProtocolError> {
    match value {
        Value::Null => output.extend_from_slice(b"null"),
        Value::Bool(true) => output.extend_from_slice(b"true"),
        Value::Bool(false) => output.extend_from_slice(b"false"),
        Value::String(value) => output.extend_from_slice(
            serde_json::to_string(value)
                .map_err(|error| ProtocolError::new(error.to_string()))?
                .as_bytes(),
        ),
        Value::Number(number) => {
            let encoded = if let Some(value) = number.as_u64() {
                if value > MAX_SAFE_JSON_INTEGER {
                    return Err(ProtocolError::new(
                        "proof JSON integer exceeds the portable safe range",
                    ));
                }
                value.to_string()
            } else if let Some(value) = number.as_i64() {
                if value.unsigned_abs() > MAX_SAFE_JSON_INTEGER {
                    return Err(ProtocolError::new(
                        "proof JSON integer exceeds the portable safe range",
                    ));
                }
                value.to_string()
            } else {
                return Err(ProtocolError::new(
                    "floating-point values are forbidden in proof JSON",
                ));
            };
            output.extend_from_slice(encoded.as_bytes());
        }
        Value::Array(values) => {
            output.push(b'[');
            for (index, value) in values.iter().enumerate() {
                if index > 0 {
                    output.push(b',');
                }
                write_canonical(value, output)?;
            }
            output.push(b']');
        }
        Value::Object(values) => {
            let mut entries = values.iter().collect::<Vec<_>>();
            entries.sort_by(|(left, _), (right, _)| left.encode_utf16().cmp(right.encode_utf16()));
            output.push(b'{');
            for (index, (key, value)) in entries.into_iter().enumerate() {
                if index > 0 {
                    output.push(b',');
                }
                output.extend_from_slice(
                    serde_json::to_string(key)
                        .map_err(|error| ProtocolError::new(error.to_string()))?
                        .as_bytes(),
                );
                output.push(b':');
                write_canonical(value, output)?;
            }
            output.push(b'}');
        }
    }
    Ok(())
}
